javatests/arcs/android/storage/database/DatabaseImplTest.kt

```diff
@@ -2837,6 +2837,525 @@ class DatabaseImplTest {
         assertThat(testClient.deleteWasCalled).isTrue()
     }
 
+    @Test
+    fun canUpgradeSchema_byAddingNewFields() = runBlockingTest {
+        val collectionKey = DummyStorageKey("collection")
+        val backingKey1 = DummyStorageKey("backing1")
+        val backingKey2 = DummyStorageKey("backing2")
+
+        val entity1 = DatabaseData.Entity(
+            RawEntity(
+                "entity1",
+                mapOf("text" to "forty two".toReferencable()),
+                emptyMap()
+            ),
+            PRE_UPGRADE_SCHEMA,
+            FIRST_VERSION_NUMBER,
+            VERSION_MAP
+        )
+
+        val collection = DatabaseData.Collection(
+            values = setOf(
+                ReferenceWithVersion(
+                    Reference("entity1", backingKey1, VersionMap("entity1" to 1)),
+                    VersionMap("actor" to 1)
+                )
+            ),
+            schema = PRE_UPGRADE_SCHEMA,
+            databaseVersion = 1,
+            versionMap = VersionMap("actor" to 1)
+        )
```

Comment:
do you need these collections, or can you just write and read entities (to simplify the tests)? I think the collection schema is irrelevant here. I also wonder if storing entity2 matters. So I would simplify this test (and similarly the others) to something like:

val entity1 = DatabaseData.Entity(
            RawEntity(
                "entity1",
                mapOf("text" to "forty two".toReferencable()),
                emptyMap()
            ),
            PRE_UPGRADE_SCHEMA,
            FIRST_VERSION_NUMBER,
            VERSION_MAP
        )
database.insertOrUpdate(backingKey1, entity1)
val readBack = database.getEntity(it.reference.storageKey, POST_UPGRADE_SCHEMA)
assertThat(readBack).isEqualTo(
// The existing entity is returned as if it belongs to the new schema,
        // but with a null value for the new field "number".
        DatabaseData.Entity(
            RawEntity(
                "entity1",
                mapOf("text" to "forty two".toReferencable(), "number" to null),
                emptyMap()
            ),
            POST_UPGRADE_SCHEMA,
            FIRST_VERSION_NUMBER,
            VERSION_MAP
        )
)

it will make more obvious what you care about in this test